### PR TITLE
Update NFT to accept compressed hashes from bridge

### PR DESCRIPTION
**Pull request by lucasvo**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 out
+keystore
```
Comment (lucasvo) [inline .gitignore]: I don't think this should go into the upstream .gitignore. You can modify your own global gitignore here: https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

File: addresses.json [hunk 2]
```diff
@@ -0,0 +1,1313 @@
+{
```
Comment (lucasvo) [inline addresses.json]: @vedhavyas this should not be comitted

File: deploy [hunk 3]
```diff
@@ -2,21 +2,35 @@
...
 
 "
 set -ex
 
-REGISTRY=$(seth send --create out/NFT.bin 'NFT(string memory, string memory, address)' "Centrifuge NFT" "NFT" $ANCHOR_REPOSITORY)
+export CONTRACT_NAME=${CONTRACT_NAME:-NFT}
+
+export CENT_ENV=${CENT_ENV:-"fulvous"}
+export ETH_GAS=${ETH_GAS:-4712388}
+export ETH_KEYSTORE=${ETH_KEYSTORE:-./keystore}
```
Comment (lucasvo) [inline deploy]: These are default gas options that shouldn't be in there

File: src/asset_verifier.sol [hunk 4]
```diff
@@ -0,0 +1,24 @@
+pragma solidity >=0.4.24;
```
Comment (lucasvo) [inline src/asset_verifier.sol]: This is an extremely old solidity version

File: src/asset_verifier.sol [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+contract IAssetManager {
+  function isAssetValid(bytes32 asset) external view returns (bool);
+}
+
+contract AssetVerifier {
```
Comment (lucasvo) [inline src/asset_verifier.sol]: Let's inline this to NFT. KISS.

File: src/nft.sol [hunk 6]
```diff
@@ -14,184 +14,29 @@
...
-    function tokenURI( uint token_id) external view returns (string memory) {
-        return string(abi.encodePacked(uri, "0x", uintToHexStr(uint(address(this))), "/0x", uintToHexStr(token_id)));
-    }
-
-  /**
-   * @dev Checks if the document is the latest version anchored
-   * @param data_root bytes32 hash of all data fields of the document which are signed
-   * @param next_anchor_id uint the next id to be anchored
-   */
-  function _latestDoc( bytes32 data_root, uint next_anchor_id)  internal view returns (bool) {
```
Comment (lucasvo) [inline src/nft.sol:148]: why was all of this removed?

File: src/asset_verifier.sol (hunk 4 above)
Comment (lucasvo) [reply src/asset_verifier.sol]: Let's update to 0.5.15 (used in tinlake as well)

File: src/asset_verifier.sol (hunk 5 above)
Comment (vedhavyas) [reply src/asset_verifier.sol]: This was inlined last time I pushed them. Shall I move this back @mikiquantum ?

File: deploy (hunk 3 above)
Comment (vedhavyas) [reply deploy]: This is deployment script, do we need these defaults @mikiquantum ?

File: src/nft.sol (hunk 6 above)
Comment (vedhavyas) [reply src/nft.sol:148]: we dont need these anymore since these no anchor repo

File: src/nft.sol [hunk 7]
```diff
@@ -14,184 +14,29 @@
...
-        return next_merkle_root_ == 0x0;
-  }
-
-  /**
-   * @dev Checks that provided document is signed by the given identity and validates and checks if the public key used is a valid SIGNING_KEY. Does not check if the signature root is part of the document root.
-   * @param anchor uint anchor ID
-   * @param data_root bytes32 hash of all data fields of the document which are signed
-   * @param signature bytes The signature used to sign the data root
-   */
-    function _signed(uint anchor, bytes32 data_root, bytes memory signature) internal view {
```
Comment (lucasvo) [inline src/nft.sol:159]: @vedhavyas this should not be deleted.

File: src/nft.sol [hunk 8]
```diff
@@ -14,184 +14,29 @@
...
-      if (revokedAt_ > 0) {
-        require(anchored_block < revokedAt_,"Document signed with a revoked key.");
-      }
-    }
-
-  /**
-   * @dev Checks that the passed in token proof matches the data for minting
-   * @param tkn uint The ID for the token to be minted
-   */
-    function _checkTokenData(uint tkn, bytes memory property, bytes memory value) internal view returns (bool) {
```
Comment (lucasvo) [inline src/nft.sol:180]: @vedhavyas this should not be deleted

File: src/nft.sol [hunk 9]
```diff
@@ -14,184 +14,29 @@
...
-
-  /**
-   * @dev Checks that provided document is signed by the given identity and validates and checks if the public key used is a valid SIGNING_KEY. Does not check if the signature root is part of the document root.
-   * @param anchor uint anchor ID
-   * @param data_root bytes32 hash of all data fields of the document which are signed
-   * @param signature bytes The signature used to sign the data root
-   */
-    function _signed(uint anchor, bytes32 data_root, bytes memory signature) internal view {
-      // Get anchored block from anchor ID
-    (, , uint32 anchored_block) = anchors.getAnchorById(anchor);
```
Comment (lucasvo) [inline src/nft.sol:161]: Given we don't have the anchored block anymore on ethereum I propose we check that the signature is presently valid.

File: src/test/nft.t.sol [hunk 10]
```diff
@@ -16,274 +16,96 @@
...
-contract AnchorMock {
-    bytes32 documentRoot;
-    uint32  blockNumber;
+contract MockAssetManager {
+    mapping (bytes32 => uint8) public assets;
 
-    function file(bytes32 documentRoot_, uint32 blockNumber_) public {
-        documentRoot = documentRoot_;
-        blockNumber = blockNumber_;
+    function store(bytes32 asset) public {
```
Comment (lucasvo) [inline src/test/nft.t.sol]: We use `file` throughout our codebase, please keep this convention unless it's implementing an specific interface.

Comment (vedhavyas) [reply src/test/nft.t.sol]: What does `file` mean here?
I'm not familiar with solidity convictions we are following, is there any document I should take a look at @lucasvo ?

File: src/nft.sol [hunk 11]
```diff
@@ -13,185 +13,156 @@
...
+pragma experimental ABIEncoderV2;
 
 import { ERC721Metadata } from "./openzeppelin-solidity/token/ERC721/ERC721Metadata.sol";
 import "./openzeppelin-solidity/cryptography/ECDSA.sol";
-import "./merkle.sol";
 
-contract AnchorLike {
-    function getAnchorById(uint) public view returns (uint, bytes32, uint32);
+contract AssetManagerLike {
+    function isAssetValid(bytes32 asset) external view returns (bool);
```
Comment (lucasvo) [inline src/nft.sol]: Why not have a method `exists()` that reverts if it does not as opposed to having to check for a return value?

File: src/nft.sol [hunk 12]
```diff
@@ -13,185 +13,156 @@
...
-            return true;
+    /**
+     * Computes the asset hash from the given params and verifies if its present in the AssetManager
+     * @param to address deposit address of the token
+     * @param properties bytes[] properties of the each proof
+     * @param values bytes[] value associated with each property
+     * @param salts bytes[] salt associated with each property
+     */
+    function _verifyAsset(address to, bytes[] memory properties, bytes[] memory values, bytes32[] memory salts) internal view returns (bool) {
+        require(to != address(0), "not a valid address");
```
Comment (lucasvo) [inline src/nft.sol]: I don't think we need to catch this use case on chain. It will fail at a later point anyway.

File: src/nft.sol [hunk 13]
```diff
@@ -13,185 +13,156 @@
...
+        // Extract the public key and identity address from the signature
+        address identity_ = data_root.toEthSignedMessageHash().recover(signature);
+        bytes32 pbKey_ = bytes32(uint(identity_) << 96);
+        // check that the identity being used has been created by the Centrifuge Identity Factory contract
+        require(identityFactory.createdIdentity(identity_), "Identity is not registered.");
+        // check that public key has signature purpose on provided identity
+        require(keyManager.keyHasPurpose(pbKey_, SIGNING_PURPOSE), "Signature key is not valid.");
     }
 
-    function tokenURI( uint token_id) external view returns (string memory) {
```
Comment (lucasvo) [inline src/nft.sol:139]: Why was this removed?

File: src/nft.sol (hunk 11 above)
Comment (lucasvo) [reply src/nft.sol]: I think the basic AssetRegistry implementation should look like this:
```
contract AssetRegistry {
    mapping (bytes32 => bool) public assets;
    function exists(bytes32 asset) public view {
        require(assets[asset], "asset-registry/not-found");
    }
}
```
Therefore the Abstract can look like this:

```
contract AbstractRegistryLike {
    function assets(bytes32) public view returns (bool);
    function exists(bytes32) public view;
}
```
This makes a check very simple:
```
registry.exists(asset);
```

File: src/nft.sol [hunk 14]
```diff
@@ -13,185 +13,156 @@
...
-        } else {
-           return doc_root == sha256(concat(sig_root, data_root));
-        }
+    /**
+     * @dev Checks that provided document is signed by the given identity. Validates and checks if the public key used is a valid SIGNING_KEY.
+     * Does not check if the signature root is part of the document root.
+     * @param data_root bytes32 hash of all data fields of the document which are signed
+     * @param signature bytes The signature used to sign the data root
+     */
+    function _signed(bytes32 data_root, bytes memory signature) internal view {
```
Comment (lucasvo) [inline src/nft.sol]: This function only verifies that the signature is made by **any** key known. We should add an argument `address identity`.
```suggestion
    function _signed(bytes32 identity, bytes32 data_root, bytes memory signature) internal view {
```

File: src/nft.sol [hunk 15]
```diff
@@ -13,185 +13,156 @@
...
+        // Extract the public key and identity address from the signature
+        address identity_ = data_root.toEthSignedMessageHash().recover(signature);
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
        address key_ = data_root.toEthSignedMessageHash().recover(signature);
```

File: src/nft.sol [hunk 16]
```diff
@@ -13,185 +13,156 @@
...
+        bytes32 pbKey_ = bytes32(uint(identity_) << 96);
+        // check that the identity being used has been created by the Centrifuge Identity Factory contract
+        require(identityFactory.createdIdentity(identity_), "Identity is not registered.");
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
        require(identityFactory.createdIdentity(identity), "Identity is not registered.");
```

File: src/nft.sol [hunk 17]
```diff
@@ -13,185 +13,156 @@
...
-        key_manager = KeyManagerLike(identity_);
-        identity_factory = IdentityFactoryLike(identity_factory_);
+    AssetManagerLike assetManager;
+    IdentityFactoryLike identityFactory;
+    KeyManagerLike keyManager;
+
+    constructor(string memory name, string memory symbol, address assetManager_, address identityFactory_, address keyManager_) ERC721Metadata(name, symbol) public {
+        assetManager = AssetManagerLike(assetManager_);
+        identityFactory = IdentityFactoryLike(identityFactory_);
+        keyManager = KeyManagerLike(keyManager_);
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
```

File: src/nft.sol [hunk 18]
```diff
@@ -13,185 +13,156 @@
...
 
-    constructor (string memory name, string memory symbol, address anchors_, address identity_, address identity_factory_) ERC721Metadata(name, symbol) public {
-        anchors = AnchorLike(anchors_);
-        key_manager = KeyManagerLike(identity_);
-        identity_factory = IdentityFactoryLike(identity_factory_);
+    AssetManagerLike assetManager;
+    IdentityFactoryLike identityFactory;
+    KeyManagerLike keyManager;
+
+    constructor(string memory name, string memory symbol, address assetManager_, address identityFactory_, address keyManager_) ERC721Metadata(name, symbol) public {
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
    constructor(string memory name, string memory symbol, address assetManager_, address identityFactory_) ERC721Metadata(name, symbol) public {
```

File: src/nft.sol [hunk 19]
```diff
@@ -13,185 +13,156 @@
...
+     * @param data_root bytes32 hash of all data fields of the document which are signed
+     * @param signature bytes The signature used to sign the data root
+     */
+    function _signed(bytes32 data_root, bytes memory signature) internal view {
+        // Extract the public key and identity address from the signature
+        address identity_ = data_root.toEthSignedMessageHash().recover(signature);
+        bytes32 pbKey_ = bytes32(uint(identity_) << 96);
+        // check that the identity being used has been created by the Centrifuge Identity Factory contract
+        require(identityFactory.createdIdentity(identity_), "Identity is not registered.");
+        // check that public key has signature purpose on provided identity
```
Comment (lucasvo) [inline src/nft.sol:106]:
```suggestion
        // check that public key has signature purpose on provided identity
        KeyManager keyManager = KeyManager(identity);
```

File: src/nft.sol [hunk 20]
```diff
@@ -13,185 +13,156 @@
...
+        require(keyManager.keyHasPurpose(pbKey_, SIGNING_PURPOSE), "Signature key is not valid.");
```
Comment (lucasvo) [inline src/nft.sol]: Checking the `keyHasPurpose` method reveals that it does not verify if the key is revoked or not: 
https://github.com/centrifuge/centrifuge-ethereum-contracts/blob/develop/contracts/KeyManager.sol#L151

Instead we need to use 'getKey' https://github.com/centrifuge/centrifuge-ethereum-contracts/blob/develop/contracts/KeyManager.sol#L130

File: src/nft.sol (hunk 15 above)
Comment (mikiquantum) [reply src/nft.sol]: Since we append the transition flag to the data_root/signing_root then this recover will not work, it will never recover the same key.
Some time back I created a custom `toEthSignedMessageHash().recover()` function that needs to be used/moved here. Here is the code snippet: 
https://github.com/centrifuge/centrifuge-ethereum-contracts/commit/05ea88324367e29eb2b5bd872b2d9819cd3c9cbc#diff-52561fd7440284be9f029c743cdcf837R111-R133

File: src/nft.sol (hunk 20 above)
Comment (mikiquantum) [reply src/nft.sol]: Yep, with `found_key.revokedAt == 0`

File: src/nft.sol (hunk 11 above)
Comment (mikiquantum) [reply src/nft.sol]: I think **read** operations should return a value or not. I believe the responsibility to react with an exception should be done outside of the read function itself.

Comment (vedhavyas) [reply src/nft.sol]: @lucasvo wdyt ?

Comment (lucasvo) [reply src/nft.sol]: No, this is a commonly used pattern and leads to less code. We use this pattern in many places in our tinlake contracts. I've renamed the method to `verify` make it clear it is not used to read a value but simply verifies the asset.

```
contract AssetRegistry {
    mapping (bytes32 => bool) public assets;
    function verify(bytes32 asset) public view {
        require(assets[asset], "asset-registry/not-found");
    }
}

contract AbstractRegistryLike {
    function assets(bytes32) public view returns (bool);
    function verify(bytes32) public view;
}
```

Comment (lucasvo) [reply src/nft.sol]: Solidity auto generates a read function already by having a `public` mapping:
```
    mapping (bytes32 => bool) public assets;
```

The solidity compiles this to the same bytecode as:
```
contract AssetRegistry {
    mapping (bytes32 => bool) assets;
    function assets(bytes32 item) public view returns (bool) {
        return assets[item];
    }
}
```
Writing a custom getter is redundant.

File: src/nft.sol (hunk 12 above)
Comment (lucasvo) [reply src/nft.sol]: I've reopened this until it is addressed in the PR

Comment (vedhavyas) [reply src/nft.sol]: always fail-fast :)

Comment (vedhavyas) [reply src/nft.sol]: I have removed this check for now, but I would love it be there

Comment (lucasvo) [reply src/nft.sol]: fail fast for extra gas is not worth it IMO. 99.999% of times this will be not be triggered but it will still cost people gas every time.

File: src/nft.sol [hunk 21]
```diff
@@ -82,23 +83,45 @@ contract NFT is ERC721Metadata {
             data = abi.encodePacked(data, keccak256(abi.encodePacked(properties[i], values[i], salts[i])));
         }
 
-        return assetManager.isAssetValid(keccak256(data));
+        assetManager.verify(keccak256(data));
     }
 
     /**
-     * @dev Checks that provided document is signed by the given identity. Validates and checks if the public key used is a valid SIGNING_KEY.
+     * @dev Checks that provided document is signed by the given identity. Validates and checks if the public key used is a not revoked.
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
     * @dev Checks that provided document is signed by the given identity. Validates and checks if the public key used is a not revoked and has the purpose SIGNING_KEY.
```

File: src/nft.sol [hunk 22]
```diff
@@ -82,23 +83,45 @@ contract NFT is ERC721Metadata {
...
-     * @param signature bytes The signature used to sign the data root
+     * @param identity address identity which is a collaborator of the document
+     * @param dataRoot bytes32 hash of all data fields of the document which are signed
+     * @param signature bytes contains signature + transition flag
      */
-    function _signed(bytes32 data_root, bytes memory signature) internal view {
-        // Extract the public key and identity address from the signature
-        address identity_ = data_root.toEthSignedMessageHash().recover(signature);
-        bytes32 pbKey_ = bytes32(uint(identity_) << 96);
+    function _signed(address identity, bytes32 dataRoot, bytes memory signature) internal view {
```
Comment (lucasvo) [inline src/nft.sol]: Maybe we should name this signed_document instead as the method always verifies our custom document signatures that sign dataroot+transition flag.
```suggestion
    function _signed_document(address identity, bytes32 dataRoot, bytes memory signature) internal view {
```

File: src/nft.sol [hunk 23]
```diff
@@ -82,23 +83,45 @@ contract NFT is ERC721Metadata {
...
         // check that the identity being used has been created by the Centrifuge Identity Factory contract
-        require(identityFactory.createdIdentity(identity_), "Identity is not registered.");
+        require(identityFactory.createdIdentity(identity), "Identity is not registered.");
+        // extract flag from signature
+        (bytes memory sig, byte flag) = _recoverSignatureAndFlag(signature);
+        // Recalculate hash and extract the public key from the signature
+        address key = _reCalculateHash(dataRoot, flag).recover(sig);
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
        address key = _toEthSignedMessage(dataRoot, flag).recover(sig);
```

File: src/nft.sol [hunk 24]
```diff
@@ -82,23 +83,45 @@ contract NFT is ERC721Metadata {
...
+        bytes32 pubKey = bytes32(uint(key) << 96);
         // check that public key has signature purpose on provided identity
-        require(keyManager.keyHasPurpose(pbKey_, SIGNING_PURPOSE), "Signature key is not valid.");
+        KeyManagerLike keyManager = KeyManagerLike(identity);
+        require(keyManager.keyHasPurpose(pubKey, SIGNING_PURPOSE), "Signature key is not valid.");
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
        require(keyManager.keyHasPurpose(pubKey, SIGNING_PURPOSE), "nft/sig-key-not-valid");
```

File: src/nft.sol [hunk 25]
```diff
@@ -82,23 +83,45 @@ contract NFT is ERC721Metadata {
...
+        (, , uint32 revokedAt) = keyManager.getKey(pubKey);
+        require(revokedAt == 0, "key is revoked");
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
        require(revokedAt == 0, "nft/key-revoked");
```

File: src/nft.sol [hunk 26]
```diff
@@ -82,23 +83,45 @@ contract NFT is ERC721Metadata {
...
+     * @param signature bytes contains signature + transition flag
      */
-    function _signed(bytes32 data_root, bytes memory signature) internal view {
-        // Extract the public key and identity address from the signature
-        address identity_ = data_root.toEthSignedMessageHash().recover(signature);
-        bytes32 pbKey_ = bytes32(uint(identity_) << 96);
+    function _signed(address identity, bytes32 dataRoot, bytes memory signature) internal view {
         // check that the identity being used has been created by the Centrifuge Identity Factory contract
-        require(identityFactory.createdIdentity(identity_), "Identity is not registered.");
+        require(identityFactory.createdIdentity(identity), "Identity is not registered.");
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
        require(identityFactory.createdIdentity(identity), "nft/identity-not-registered");
```

File: src/nft.sol [hunk 27]
```diff
@@ -82,23 +83,45 @@ contract NFT is ERC721Metadata {
...
-        require(keyManager.keyHasPurpose(pbKey_, SIGNING_PURPOSE), "Signature key is not valid.");
+        KeyManagerLike keyManager = KeyManagerLike(identity);
+        require(keyManager.keyHasPurpose(pubKey, SIGNING_PURPOSE), "Signature key is not valid.");
+        (, , uint32 revokedAt) = keyManager.getKey(pubKey);
+        require(revokedAt == 0, "key is revoked");
+    }
+
+    function _recoverSignatureAndFlag(bytes memory signature) internal pure returns (bytes memory, byte) {
+        // ensure signature value is 66
+        require(signature.length == 66, "not a valid signature value");
```
Comment (lucasvo) [inline src/nft.sol]:
```suggestion
        require(signature.length == 66, "nft/invalid-signature-length");
```

File: src/nft.sol [hunk 28]
```diff
@@ -32,166 +32,163 @@ contract IdentityFactoryLike {
...
+     * @param signature bytes contains signature + transition flag
+     */
+    function _signed_document(address identity, bytes32 dataRoot, bytes memory signature) internal view {
+        // check that the identity being used has been created by the Centrifuge Identity Factory contract
+        require(identityFactory.createdIdentity(identity), "nft/identity-not-registered");
+        // extract flag from signature
+        (bytes memory sig, byte flag) = _recoverSignatureAndFlag(signature);
+        // Recalculate hash and extract the public key from the signature
+        address key = _toEthSignedMessage(dataRoot, flag).recover(sig);
+        bytes32 pubKey = bytes32(uint(key) << 96);
```
Comment (xmxanuel) [inline src/nft.sol:105]:
```suggestion
// address has 20 bytes
// pubkey has 32 bytes we bit shift (12*8=96) to the left
        bytes32 pubKey = bytes32(uint(key) << 96);
```

File: src/nft.sol [hunk 29]
```diff
@@ -32,166 +32,163 @@ contract IdentityFactoryLike {
...
-           return doc_root == sha256(concat(data_root, sig_root));
-        } else {
-           return doc_root == sha256(concat(sig_root, data_root));
+    function _recoverSignatureAndFlag(bytes memory signature) internal pure returns (bytes memory, byte) {
+        // ensure signature value is 66
+        require(signature.length == 66, "nft/invalid-signature-length");
+        byte flag = signature[65];
+        bytes memory sig = new bytes(65);
+        uint i = 0;
+        for (i = 0; i < 65; i++) {
```
Comment (xmxanuel) [inline src/nft.sol:119]: can you optimize this?
https://stackoverflow.com/questions/49051856/is-there-a-pop-functionality-for-solidity-arrays

try:
```javascript
delete myArray[myArray.length - 1];
myArray.length--;
```

File: src/nft.sol [hunk 30]
```diff
@@ -32,166 +32,163 @@ contract IdentityFactoryLike {
...
+     * @param salts bytes[] salt associated with each property
+     */
+    function _verifyAsset(address to, bytes[] memory properties, bytes[] memory values, bytes32[] memory salts) internal view  {
+        // construct assetHash from the props, values, salts
+        // append to address
+        bytes memory data = abi.encodePacked(to);
+
+        // append hashes
+        for (uint i=0; i< properties.length; i++) {
+            data = abi.encodePacked(data, keccak256(abi.encodePacked(properties[i], values[i], salts[i])));
```
Comment (xmxanuel) [inline src/nft.sol:83]: I was not involved in designing it. 

What's the reasoning behind this instead of a tree structure?

File: src/nft.sol [hunk 31]
```diff
@@ -32,166 +32,163 @@ contract IdentityFactoryLike {
...
+     * Does not check if the signature root is part of the document root.
+     * @param identity address identity which is a collaborator of the document
+     * @param dataRoot bytes32 hash of all data fields of the document which are signed
+     * @param signature bytes contains signature + transition flag
+     */
+    function _signed_document(address identity, bytes32 dataRoot, bytes memory signature) internal view {
+        // check that the identity being used has been created by the Centrifuge Identity Factory contract
+        require(identityFactory.createdIdentity(identity), "nft/identity-not-registered");
+        // extract flag from signature
+        (bytes memory sig, byte flag) = _recoverSignatureAndFlag(signature);
```
Comment (xmxanuel) [inline src/nft.sol:100]: why are you not passing the `flag` as an extra parameter?

What's the benefit in splitting the flag from the signature on-chain?

File: src/nft.sol (hunk 30 above)
Comment (lucasvo) [reply src/nft.sol:83]: because you always need to proof all values anyway so a tree is actually less efficient.

File: src/nft.sol (hunk 31 above)
Comment (lucasvo) [reply src/nft.sol:100]: Because it is encoded in the signature leaf of the document already.

File: src/nft.sol (hunk 29 above)
Comment (vedhavyas) [reply src/nft.sol:119]: I have tried this before. This would not work since memory array cannot be resized. And another option is to have the array as storage but that would cost gas. Hence, this for loop @manuelpolzhofer

Comment (vedhavyas) [reply src/nft.sol:119]: We can go with assembly route, I'm not very proficient on that, but still, doubt its worthiness regarding the optimisations
